### ISSUE TITLE
Request for public APIs or alternatives: TF-Addons

### ISSUE THREAD

**Issue by seanpmorgan** [tensorflow/tensorflow#28601]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): TF 2.0

As part of our effort to make [tensorflow/addons](https://github.com/tensorflow/addons) independent of internal TensorFlow core APIs we've identified several things that we cannot find public API alternatives for; or would like to discuss other options. One such option is to copy the code statically into addons, but for anything that is likely to change in the future this isn't great.

cc @karmel 

## Testing
We heavily depend on testing decorators that are not exposed in the public API.

* [tensorflow.python.framework.test_util.run_all_in_graph_and_eager_modes](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/test_util.py#L893)
* [tensorflow.python.framework.test_util.run_in_graph_and_eager_modes](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/test_util.py#L957)
* [tensorflow.python.keras.keras_parameterized.run_all_keras_modes](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/keras_parameterized.py#L176)
* [tensorflow.python.keras.testing_utils.layer_test](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/testing_utils.py#L71)
    * `layer_test` has been very helpful for catching serialization issues and other problems with our custom layers.
* [tensorflow.python.framework.test_util.run_deprecated_v1](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/test_util.py#L1128)
     * `run_deprecated_v1` can probably be statically copied over? We should be looking to move away from this anyhow.

## Keras
* [tensorflow.python.keras.losses.LossFunctionWrapper](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/losses.py#L177)
     * `LossFunctionWrapper` is used to convert our loss functions to a Keras Loss. It seems extensively used in core repository, but there is no API. Should we be refactoring our functions to inherit from base Loss? And if so why not do that internally as well?
* [tensorflow.python.keras.engine.base_layer_utils.mark_checked](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/base_layer_utils.py#L414)
     * This looks like we could easily re-create it using tf.nest. Just want to confirm that this isn't expected to used outside of TF often.
 

## RNN Ops
* [tensorflow.python.ops. rnn._transpose_batch_time](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn.py#L43)
* [tensorflow.python.ops. rnn_cell_impl.assert_like_rnncell](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell_impl.py#L75)
* [tensorflow.python.ops. rnn_cell_impl._zero_state_tensors](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell_impl.py#L171)

## Control Flow
* [tensorflow.python.ops.control_flow_util.GetContainingXLAContext](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/control_flow_util.py#L200)
* [tensorflow.python.ops.control_flow_util.GetContainingWhileContext](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/control_flow_util.py#L178)

**Comment by karmel** [comment]
Thanks @seanpmorgan . Will review with the API owners.

cc @fchollet for the Keras methods, and cc @qlzh727 for the RNN methods, in case you have any opinions about whether these should/can be made public.

**Comment by qlzh727** [comment]
I will take look for the RNN one, I might have missed them during rewrite for the seq2seq. The control flow one might also be in similar case.

**Comment by alextp** [comment]
For the ones below

```
tensorflow.python.framework.test_util.run_all_in_graph_and_eager_modes
tensorflow.python.framework.test_util.run_in_graph_and_eager_modes
tensorflow.python.keras.keras_parameterized.run_all_keras_modes
tensorflow.python.framework.test_util.run_deprecated_v1
```

@isaprykin has been working on a more reliable testing infrastructure for tf2 that will allow finer grained control over the combinations. I'd say wait for that.

For `tensorflow.python.keras.testing_utils.layer_test` I think you should fork  it and expose it. It mostly only depends on public tf symbols (and we need to make the cudnn deterministic one public) but it has no reason to be subject to TF's stability guarantees and release cadence. I think it'd be a seriously useful tool which third-party layer writer will get by importing addons.

Re LossFunctionWrapper I encourage you to inherit from Loss instead.

transpose_batch_time you should fork.

Re the XLA context one and the while context one I'd really really like to make it so you don't have any need for those things to exist. Can we discuss separately why do you need them?

**Comment by qlzh727** [comment]
I think the XLA context was in the seq2seq code https://github.com/tensorflow/tensorflow/blob/e4371880b1a50f7b23b2375c9c7c7fcd2dcf1a5b/tensorflow/contrib/seq2seq/python/ops/decoder.py#L189. I didn't remove them during the rewrite since I don't want to break XLA when we port the addon code back to third_party/tensorflow_addons.

**Comment by Squadrick** [comment]
Can we expose [`tf.keras.metrics.MeanMetricWrapper`](https://github.com/tensorflow/tensorflow/blob/18e619cc61952cb89863dc2a8d5338bd321453ef/tensorflow/python/keras/metrics.py#L526) to the public API? We're implementing new metrics in TensorFlow Addons, and some of them are stateless functions that can be easily wrapped in `MeanMetricWrapper`.

**Comment by karmel** [comment]
@brainnoise -- thoughts on MeanMetricWrapper?

**Comment by seanpmorgan** [comment]
> @brainnoise -- thoughts on MeanMetricWrapper?

Just confirming this is the correct username? Looks like it may have been deleted?

**Comment by qlzh727** [comment]
@pavithrasv

**Comment by Squadrick** [comment]
@karmel @qlzh727 @pavithrasv Any update on this?

**Comment by pavithrasv** [comment]
`MeanMetricWrapper` is an internal helper class to be able to create `Metric` class wrapper around stateless metric functions. This is not required for use with any Keras API. eg., you can pass a stateless fn `def my_metric(y_true, y_pred)` to the Keras compile API as is. 

Because of this i am leaning towards having a version of the `MeanMetricWrapper` in the Add ons repo may be?

**Comment by ematejska** [comment]
The API owners are moving to a rotation based system instead of the meetings. Adding a couple of owners to follow along.

**Comment by bhack** [comment]
Just to notify that the related RFC was just claimed closed/implemented:

https://github.com/keras-team/keras-cv/blob/master/setup.py#L29

**Comment by qlzh727** [comment]
@bhack, do u still have a list of the function/private API that are used by addon? I can revisit them and see if we will rewrite or expose test related __intenral__ APIs.

**Comment by bhack** [comment]
From a quick check on the TFA checkout `grep -R  "from tensorflow.python.keras" --include=*.py`:
```
tensorflow_addons/utils/types.py:    from tensorflow.python.keras.engine import keras_tensor
tensorflow_addons/utils/test_utils.py:    from tensorflow.python.keras.testing_utils import layer_test  # noqa: F401
tensorflow_addons/seq2seq/attention_wrapper.py:from tensorflow.python.keras.engine import base_layer_utils

```

**Comment by bhack** [comment]
Also a little bit unrelated but we had to replicate code for:
https://github.com/tensorflow/tensorflow/pull/45742

But I don't know if it will be still a problem in the next months if you want to migrate this ops in keras cv.

**Comment by qlzh727** [comment]
Trying to address one of the import in https://github.com/tensorflow/addons/pull/2655.

For KerasTensor type check, I don't think we are going to expose is it as any public API soon(although it is a return value from a public API), I will need to check with API owner @fchollet for this. I don't think it will be too bad if we remove that in the tensorlike list.

Similar for layer_test, since it is a testing util (and keras hasn't historically expose as test util), I will need API owner @fchollet  to approve this change as well (decide an API namespace).

**Comment by sushreebarsa** [comment]
@seanpmorgan Could you please refer to the [comment](https://github.com/tensorflow/tensorflow/issues/28601#issuecomment-1035599709) above and let us know if it helps?Thanks!

**Comment by bhack** [comment]
@seanpmorgan I think that as we cannot have `keras_tensor` exposed this could be closed.
